src/fhda/tortoise.py

```diff
@@ -0,0 +1,256 @@
+from futurehouse_client.models.app import AuthType
+
+
+class StepConfig(BaseModel):
+    """Configuration for a step in the pipeline"""
+
+    language: str = Field(
+        default="PYTHON", description="Language for execution environment"
+    )
+    max_steps: int = Field(
+        default=30, description="Maximum number of steps for the agent"
+    )
+    timeout: int = Field(default=15 * 60, description="Timeout for the step in seconds")
+    eval: bool = Field(default=True, description="Whether to use eval mode")
+
+
+class Step(BaseModel):
+    """A step in the pipeline"""
+
+    name: str = Field(
+        description="Name of the job to run (e.g. 'job-futurehouse-data-analysis-crow-high')"
+    )
+    prompt: str = Field(description="Prompt template to use for the step")
+    cot_prompt: bool = Field(
+        default=False, description="Whether to augment the query with COT prompting"
+    )
+    prompt_args: Dict[str, Any] = Field(
+        default_factory=dict, description="Arguments to format the prompt template"
+    )
+    input_files: Dict[str, str] = Field(
+        default_factory=dict, description="Files to upload {'source_path': 'dest_name'}"
+    )
+    output_files: Dict[str, str] = Field(
+        default_factory=dict,
+        description="Files to download {'source_name': 'dest_path'}",
+    )
+    step_id: str = Field(
+        default_factory=lambda: str(uuid.uuid4())[:8],
+        description="Small UID for the step",
+    )
+    upload_id: Optional[str] = Field(default=None, description="Upload ID for GCS")
+    parallel: int = Field(default=1, description="Number of parallel tasks to run")
+    config: StepConfig = Field(
+        default_factory=StepConfig, description="Configuration for the step"
+    )
+    post_process: Optional[Callable] = Field(
+        default=None, description="Function to run after step completion"
+    )
+    prompt_generator: Optional[Callable] = Field(
+        default=None,
+        description="Function to generate prompts and args for parallel tasks based on previous results",
+    )
+
+    def cot_prompting(self, query, language):
+        guidelines = prompts.GENERAL_NOTEBOOK_GUIDELINES.format(language=language)
+        if language == "R":
+            guidelines = prompts.R_SPECIFIC_GUIDELINES.format(language=language)
+        return (
+            f"{prompts.CHAIN_OF_THOUGHT_AGNOSTIC.format(language=language)}\n"
+            f"{guidelines}"
+            f"Here is the research question to address:\n"
+            f"<query>\n"
+            f"{query}\n"
+            f"</query>\n"
+        )
+
+    def format_prompt(self) -> str:
+        """Format the prompt template with the provided arguments"""
+        final_prompt = self.prompt.format(**self.prompt_args)
+        if self.cot_prompt:
+            final_prompt = self.cot_prompting(final_prompt, self.config.language)
+        return final_prompt
+
+
+class Tortoise:
+    """Framework for running multi-step agent pipelines"""
```

Comment:
```suggestion
    """Runner for multi-step agent pipelines."""
```

This looks more like a runner class than a framework to me. The entire module is the framework